Source/WebCore/Modules/reporting/ReportingScope.cpp

```diff
@@ -0,0 +1,159 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ReportingScope.h"
+
+#include "ContextDestructionObserver.h"
+#include "HeaderFieldTokenizer.h"
+#include "ReportingObserver.h"
+#include "ScriptExecutionContext.h"
+#include <wtf/text/StringParsingBuffer.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(ReportingScope);
+
+Ref<ReportingScope> ReportingScope::create(ScriptExecutionContext& scriptExecutionContext)
+{
+    return adoptRef(*new ReportingScope(scriptExecutionContext));
+}
+
+ReportingScope::ReportingScope(ScriptExecutionContext& scriptExecutionContext)
+    : ContextDestructionObserver(&scriptExecutionContext)
+{
+}
+
+ReportingScope::~ReportingScope() = default;
+
+void ReportingScope::registerReportingObserver(ReportingObserver& observer)
+{
+    if (m_reportingObservers.containsIf([&observer](const auto& item) { return item.ptr() == &observer; }))
+        return;
+    
+    m_reportingObservers.append(observer);
+}
+
+void ReportingScope::unregisterReportingObserver(ReportingObserver& observer)
+{
+    m_reportingObservers.removeFirstMatching([&observer](auto item) {
+        return item.ptr() == &observer;
+    });
+}
+
+void ReportingScope::removeAllObservers()
+{
+    m_reportingObservers.clear();
+}
+
+void ReportingScope::clearReports()
+{
+    m_queuedReports.clear();
+}
+
+void ReportingScope::notifyReportObservers(Ref<Report>&& report)
+{
+    // https://www.w3.org/TR/reporting-1/#notify-observers
+    // Step 4.2: Notify reporting observers on scope with report
+    size_t reportingObservers = m_reportingObservers.size();
+    UNUSED_VARIABLE(reportingObservers);
+
+    // Step 4.2.1
+    auto possibleReportObservers = copyToVector(m_reportingObservers);
+    for (auto& observer : possibleReportObservers)
+        observer->appendQueuedReportIfCorrectType(report);
+
+    // Step 4.2.2
+    m_queuedReports.append(WTFMove(report));
+
+    // Step 4.2.3-4: If scope’s report buffer now contains more than 100 reports with type equal to type, remove the earliest item with type equal to type in the report buffer.
+    // FIXME(244369): When we support more than one time of report, remove only the specific type of report.
+    if (m_queuedReports.size() > 100)
+        m_queuedReports.remove(0);
```

Comment:
This is rather inefficient with Vector. Why not use Deque instead?

Reply:
Good point -- I should have done that from the start.